### ISSUE TITLE
why it's so slow?

### ISSUE THREAD

**Issue by lonlonago** [Maghoumi/cudbscan#1]
I run on my GTX1060，it takes almost 2s, why?

**Comment by Maghoumi** [comment]
First, please note that this implementation is suitable for large numbers of points (a CPU-only implementation is faster for smaller number of points). I just implemented their paper without any additional optimizations.

What exactly are you expectations? The default example in `main.cpp` tries to cluster 20000 random points. As it reports on my machine (with a Ryzen 9 3900X CPU), the speed up is more than 5x:

```
[CUDA] Number of clusters: 1     Time: 1.28364
[CPU] Number of clusters: 1      Time: 6.97834
Speedup was: 5.43638
```

**Comment by lonlonago** [comment]
Thanks for your work , I read your code and the paper, seems you do not use optimizations the paper noted, such as  : Use memory coalescing, shared memory, Loop unrolling,  eps^2 and so on .

**Comment by Maghoumi** [comment]
Right, this is not a fully optimized implementation. For what I wanted to do, this was enough so I didn't bother going through the optimizations.

**Comment by lonlonago** [comment]
Thanks, got it.  

from your code, I found one problem is even without distance calculation and other steps, just transmission the data from cpu to gpu, it will cost 100ms, that's much bigger than the  total time  the paper noted.

another problem is when point number over 28000, it will report cuda memery error, because 'clusterMatrix' is too large.
